JavaScript/Next-js.md

```diff
@@ -0,0 +1,221 @@
+
+Also, any customization of webpack or babel inside CRA requires ejecting, which is the one-way operation, and from that point, you are responsible for maintaining configuration files.
+
+Soon every project would probably introduce something new so we would end up with many different projects very hard to maintain.
+Next.js solves above-stated problems and allows us to use a single stack for many different projects. Plus next.js introduces some handy features, which comes out of the box:
+
+- Code splitting
+- SSR
+- Page prefetching
+- Dynamic import of modules
+- Built-in Zero-Config TypeScript Support
+- Integrated client-side routing
+- Custom server with custom express routes
+- static HTML export
+
+## POC
+
+We made POC which used some of the above features. We wanted to build an app which could be exported to static Html files (wouldn't need any server processing) and would contain modals which we could access via link.
+
+We used this requirement since many of our projects use modals, and it proved to be very useful to access them directly by adding query params to url.
+
+### App structure
+
+```
+...
+├─ components
+⎮   ├─ higherOrderComponents
+⎮   ⎮   └─ WithAuth.tsx
+⎮   ⎮   └─ WithModal.tsx
+⎮   ├─ layouts
+⎮   ⎮   └─ Layout.tsx
+⎮   ├─ modals
+|   |   └─ index.tsx
+|   |   └─ Newsletter.tsx
+|   └─ LoginRegisterForm.tsx
+├─ pages
+|   ├─ index.tsx
+⎮   ├─ login.tsx
+⎮   ├─ register.tsx
+|   └─ with-hoc-modal.tsx
+├─ services
+|   └─ auth.ts
+...
+```
+
+`login.tsx` and `register.tsx` are public pages whose functionality is self-explanatory. Once login is successful user is redirected to the index page which acts as the home page.
+
+### Layouts
+
+`Layout.tsx` is an example of a layout page for all private pages. Here is an example of how we can extend `Head` section for adding global fonts, but more importantly, we use layout page with higher order component `WithAuth` which is responsible for authentification.
+
+```jsx
+// /components/layouts/layout.tsx
+import React, { Fragment } from 'react';
+import Head from 'next/head';
+
+import withAuth from '../higherOrderComponents/withAuth';
+
+const Layout = (props) => {
+  return (
+    <Fragment>
+      <Head>
+        <link
+          href="https://fonts.googleapis.com/css?family=Roboto"
+          rel="stylesheet"
+          key="google-font-cabin"
+        />
+      </Head>
+
+      <div id="portal"></div>
+
+      <main>
+        <div className="container">{props.children}</div>
+      </main>
+
+      <style global jsx>{`
+        body {
+          font-family: 'Roboto', sans-serif;
+        }
+      `}</style>
+    </Fragment>
+  );
+};
+
+export default withAuth(Layout);
```

Comment:
uppermost component should be wrapped with withAuth. We talked about that in case of withModal.